core/src/main/scala/akka/persistence/cassandra/journal/CassandraRecovery.scala

```diff
@@ -20,21 +25,50 @@ trait CassandraRecovery extends ActorLogging {
     fromSequenceNr: Long,
     toSequenceNr:   Long,
     max:            Long
-  )(replayCallback: (PersistentRepr) => Unit): Future[Unit] =
-    queries
-      .eventsByPersistenceId(
-        persistenceId,
-        fromSequenceNr,
-        toSequenceNr,
-        max,
-        replayMaxResultSize,
-        None,
-        "asyncReplayMessages",
-        someReadConsistency,
-        someReadRetryPolicy
-      )
+  )(replayCallback: (PersistentRepr) => Unit): Future[Unit] = {
+
+    // No other writes for this pid should be taking place during recovery
+    // Expecting the result set to be small (< 10 rows)so call to all() should be safe
+    val tagProgress: Future[Map[String, TagProgress]] = preparedSelectTagProgressForPersistenceId.map(_.bind(persistenceId))
+      .flatMap(session.selectResultSet)
+      .map(rs => rs.all().foldLeft(Map.empty[String, TagProgress]) { (acc, row) =>
+        acc + (row.getString("tag") -> TagProgress(persistenceId, row.getLong("sequence_nr"), row.getLong("tag_pid_sequence_nr")))
+      })
+
+    Source.fromFutureSource(
+      tagProgress.map(tp => {
+        log.debug("Starting recovery with tag progress: {}", tp)
+        queries
+          .eventsByPersistenceId(
+            persistenceId,
+            fromSequenceNr,
+            toSequenceNr,
+            max,
+            replayMaxResultSize,
+            None,
+            "asyncReplayMessages",
+            someReadConsistency,
+            someReadRetryPolicy
+          ).map(tpr => {
+              tpr.tags.foreach(tag => {
+                tp.get(tag) match {
+                  case None =>
+                    log.debug("Tag write not in progress. Sending to TagWriter. Tag {} Sequence Nr {}.", tag, tpr.pr.sequenceNr)
+                    self ! TagWrites(tag, Vector(serializeEvent(tpr.pr, tpr.tags)))
+                  case Some(progress) =>
+                    if (tpr.pr.sequenceNr > progress.sequenceNr) {
+                      log.debug("Sequence nr > than write progress. Sending to TagWriter. Tag {} Sequence Nr {}. ", tag, tpr.pr.sequenceNr)
+                      self ! TagWrites(tag, Vector(serializeEvent(tpr.pr, tpr.tags)))
+                    }
+                }
+              })
+              tpr
+            })
+      })
+    ).map(_.pr)
       .runForeach(replayCallback)
       .map(_ => ())
+  }
```

Comment:
somewhere here it should also clear the cached tagPidSeqNr in the tag writer, for the case when persistent actor is moving back and forth between nodes

Reply:
This is done now. It's done before the recovery starts, as the recovery its self sends messages to the TagActors.